_posts/2016-02-24-session-user-flow.markdown

```diff
@@ -0,0 +1,60 @@
+---
+layout: post
+title: "Use session variables to optimize your user flow"
+date: 2016-02-24 13:37:00
+categories: ["ruby", "rails", "sessions"]
+author: "schmierkov"
+---
+
+Sessions provide you a nice little data storage feature where the application does not need to get the informations directly from the database. So you do not have to persist data in your database and can easily store infos to a user on the fly. This is a nice way to enhance the user experience on your page. Let's say you want to show all users, from a specific area, a new fancy sign up form and the rest only sees the old form. With storing these location informations in a session variable, you do not need to persist this infos in your database.
+
+<!--more-->
+
+That can be pretty handy in many ways, if you think about changing the content of a website based on infos in that session. Handling a lot of session informations can be pretty overwhelming, because you have tons of crazy methods flying around. So my idea was once to stuff all these infos into an object. This way I could test it and do wild things with the data, but the frustration comes as soon as you change your Object stored in your session.
+
+Basically this is what you should **NOT** do:
+
+```ruby
+class Location
+  def initialize(ip)
+    @location = location_request(ip)  
+  end
+
+  def city
+    @location.city
+  end
+  ...
+end
+class ApplicationController < ActionController::Base
+  before_action :set_location
+
+  def set_location
+    session[:user_location] = Location.new(user_ip)
+  end
+end
+```
```

Comment:
@schmierkov This code is missing some information. What is `user_ip`? What is `location_request`?